### PR TITLE
refactor: Make explicit CMutableTransaction -> CTransaction conversion.

### PR DESCRIPTION
**Pull request by lucash-dev**
This PR makes explicit the now implicit conversion constructor `CTransaction(const CMutableTransaction&)` in `transaction.h`. 
Minimal changes were made elsewhere to make the code compilable. I'll follow up with other PRs to address individually refactoring functions that should have a `CMutableTransaction` version, or where a `CTransaction` should be reused. 

The rationale for this change is:

 - Conversion constructors should not be explicit unless there's a strong reason for it (in the opinion of, for example, https://google.github.io/styleguide/cppguide.html, and https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ro-conversion. Let me know your take on this).
 - This particular conversion is very costly -- it implies a serialization plus hash of the transaction.
 - Even though `CTransaction` and `CMutableTransaction` represent the same data, they have very different use cases and performance properties.
 - Making it explicit allows for easier reasoning of performance trade-offs.
 - There has been previous performance issues caused by unneeded use of this implicit conversion.
 - This PR creates a map for places to look for possible refactoring and performance gains (this benefit still holds if the PR is not merged).

### CODE REVIEW

**Comment by jb55** [comment]
Concept ACK

**Review by scravy — APPROVED** [review]
utACK e4c33899e12bde83b92026a2f310333d76d87939

**Comment by promag** [comment]
> This particular conversion is very costly

Agree that for these cases an explicit constructor is preferable.

utACK e4c3389.

**Comment by MarcoFalke** [comment]
For most cases the performance should not matter (e.g. unit tests or benches where the conversion is not in a hot path. Also, for the rpc interface or tx-tools, where the overhead of communication dwarfs the speedup of skipping the conversion). For other cases, we might want to accept either a mutable or immuatable tx. See e.g.  #13359

**Comment by lucash-dev** [comment]
I agree with you that in most cases it won't make a difference. But I think having it implicit makes it harder to reason about the cases when it's important.

**Comment by MarcoFalke** [comment]
I think the only case where it could potentially matter is `src/wallet/wallet.cpp`?

**Comment by MarcoFalke** [comment]
If you wanted to have this merged first, at the very least, you'd have to reorder the two commits, since the first one doesn't compile on its own.

**Comment by lucash-dev** [comment]
@MarcoFalke 
about commit ordering: of course, will fix it.

I'll take a loot at `src/wallet/wallet.cpp` and possibly create a separate PR to be merged before this one.

I was just thinking, other changes that are just clean-up might make more sense to go as commits in this PR, instead of new PRs, since they wouldn't make that much sense to get merged if this one doesn't get merged in the end. (this doesn't apply to wallet.cpp).

I'll mark this PR as WIP meanwhile.

**Comment by lucash-dev** [comment]
Fixed commit order.

**Review by ismail120572 — APPROVED** [review]

**Comment by practicalswift** [comment]
I'm afraid this PR doesn't compile when rebase on current `master`.

**Comment by lucash-dev** [comment]
> I'm afraid this PR doesn't compile when rebase on current `master`.

That's because #14588 introduced an unnecessary implicit conversion from `CMutableTransaction` to `CTransaction` and back again.

That's precisely the kind of mistake this PR is intended to prevent.

Added commit 4f9a39b8720ba8703962341474ff74bc4be5ff48 fixing it.

File: src/bench/mempool_eviction.cpp [hunk 1]
```diff
@@ -127,7 +127,7 @@ static void MempoolEviction(benchmark::State& state)
         AddTx(tx6_r, 1100LL, pool);
         AddTx(tx7_r, 9000LL, pool);
         pool.TrimToSize(pool.DynamicMemoryUsage() * 3 / 4);
-        pool.TrimToSize(GetVirtualTransactionSize(tx1));
+        pool.TrimToSize(GetVirtualTransactionSize(CTransaction(tx1)));
```
Comment (MarcoFalke) [inline src/bench/mempool_eviction.cpp]: Should probably move out of the hot loop (see comment above) and submit as separate pull request, since it changes the benchmark

File: src/test/txvalidationcache_tests.cpp [hunk 2]
```diff
@@ -357,12 +357,12 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, TestChain100Setup)
...
         PrecomputedTransactionData txdata(tx);
         // This transaction is now invalid under segwit, because of the second input.
-        BOOST_CHECK(!CheckInputs(tx, state, pcoinsTip.get(), true, SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, true, true, txdata, nullptr));
+        BOOST_CHECK(!CheckInputs(CTransaction(tx), state, pcoinsTip.get(), true, SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, true, true, txdata, nullptr));
 
         std::vector<CScriptCheck> scriptchecks;
         // Make sure this transaction was not cached (ie because the first
         // input was valid)
-        BOOST_CHECK(CheckInputs(tx, state, pcoinsTip.get(), true, SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, true, true, txdata, &scriptchecks));
+        BOOST_CHECK(CheckInputs(CTransaction(tx), state, pcoinsTip.get(), true, SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS, true, true, txdata, &scriptchecks));
```
Comment (MarcoFalke) [inline src/test/txvalidationcache_tests.cpp]: Could submit all changes in src/test and src/wallet/test as separate pull request?

**Comment by lucash-dev** [comment]
@MarcoFalke github seems to have got lost with my latest force push, and automatically closed this PR. Should I open another one?

**Comment by fanquake** [comment]
@lucash-dev That's fine. Just point back to this PR for reference.

**Comment by lucash-dev** [comment]
Reincarnated it as #14906

**Comment by lucash-dev** [comment]
There are now two PRs (as per @MarcoFalke's request):

#14908 - Removes implicit construction of CTransaction from tests and benchmarks.
#14906 - Removes implicit construction from production code, and makes the constructor explicit.

**Comment by promag** [comment]
I think it makes sense as separate commits, not separate PRs — the changes in #14908 can only be verified with the changes in #14906 otherwise the first could be incomplete.
